packages/block-editor/src/components/writing-flow/index.js

```diff
@@ -189,324 +182,287 @@ export default function WritingFlow( { children } ) {
+				if ( isTab ) {
+					const direction = isShift ? 'findPrevious' : 'findNext';
+					// Allow tabbing between form elements rendered in a block,
+					// such as inside a placeholder. Form elements are generally
+					// meant to be UI rather than part of the content. Ideally
+					// these are not rendered in the content and perhaps in the
+					// future they can be rendered in an iframe or shadow DOM.
+					if (
+						isFormElement( target ) &&
+						isFormElement( focus.tabbable[ direction ]( target ) )
+					) {
+						return;
+					}
+
+					const next = isShift
+						? focusCaptureBeforeRef
+						: focusCaptureAfterRef;
+
+					// Disable focus capturing on the focus capture element, so it
+					// doesn't refocus this block and so it allows default behaviour
+					// (moving focus to the next tabbable element).
+					noCapture.current = true;
+					next.current.focus();
+					return;
+				} else if ( isEscape ) {
+					setNavigationMode( true );
+				}
+			}
+
+			// When presing any key other than up or down, the initial vertical
+			// position must ALWAYS be reset. The vertical position is saved so it
+			// can be restored as well as possible on sebsequent vertical arrow key
+			// presses. It may not always be possible to restore the exact same
+			// position (such as at an empty line), so it wouldn't be good to
+			// compute the position right before any vertical arrow key press.
+			if ( ! isVertical ) {
+				verticalRect = null;
+			} else if ( ! verticalRect ) {
+				verticalRect = computeCaretRect( defaultView );
+			}
 
-		// This logic inside this condition needs to be checked before
-		// the check for event.nativeEvent.defaultPrevented.
-		// The logic handles meta+a keypress and this event is default prevented
-		// by RichText.
-		if ( ! isNav ) {
-			// Set immediately before the meta+a combination can be pressed.
-			if ( isKeyboardEvent.primary( event ) ) {
-				entirelySelected.current = isEntirelySelected( target );
+			// Abort if navigation has already been handled (e.g. RichText inline
+			// boundaries).
+			if ( event.defaultPrevented ) {
+				return;
 			}
 
-			if ( isKeyboardEvent.primary( event, 'a' ) ) {
-				// When the target is contentEditable, selection will already
-				// have been set by the browser earlier in this call stack. We
-				// need check the previous result, otherwise all blocks will be
-				// selected right away.
+			if ( ! isNav ) {
 				if (
-					target.isContentEditable
-						? entirelySelected.current
-						: isEntirelySelected( target )
+					isKeyboardEvent.primary( event, 'a' ) &&
+					isEntirelySelected( target )
 				) {
 					const blocks = getBlockOrder();
 					multiSelect( first( blocks ), last( blocks ) );
 					event.preventDefault();
 				}
 
-				// After pressing primary + A we can assume isEntirelySelected is true.
-				// Calling right away isEntirelySelected after primary + A may still return false on some browsers.
-				entirelySelected.current = true;
```

Comment:
All these checks are no longer needed, we can immediately check `isEntirelySelected` now.